### PR TITLE
Write tests to check user input parser. Fix unwanted behaviour uncovered by these tests.

### PR DESCRIPTION
**Pull request by olabknbit**
@kazimierz-256 should I also write tests checking how invalid input is parsed? E.g. one action with multiple statements: `A causes f`, `A releases g`, etc?  If so, we should establish what is the desired behaviour. -> Should the parser `fail`? Should it be nondeterministic?

### CODE REVIEW

File: apps/concurrent/modules/user_input_parsing.pl [hunk 1]
```diff
@@ -52,11 +52,12 @@
     get_action_and_val(Parts, Action, Value).
 
 add_to_assoc_list("impossible", Value, Assoc, New_Assoc) :-
-    member("impossible", Assoc)
-    ->  get_assoc("impossible", Assoc, Value_Sf),
-    	append(Value_Sf, [Value], New_Value),
+    (get_assoc("impossible", Assoc, Value_Sf)
```
Comment (olabknbit) [inline apps/concurrent/modules/user_input_parsing.pl:55]: this was a bug: If there were multiple statements of type `impossible` for the same action, e.g. `impossible SHOOT at 5`, `impossible SHOOT at 7`, this was parsed to just the latter one.

Now fixed.

File: apps/concurrent/modules/user_input_parsing.pl [hunk 2]
```diff
@@ -85,11 +86,13 @@
     ;   put_assoc(Action, Assoc, Value, Domain).
 
 parse_parts([], Domain, Domain).
-parse_parts([H|T], Assoc, Domain) :-
+parse_parts([H | T], Assoc, Domain) :-
     normalize_space(atom(Line), H),
-    parse_sentence(Line, Action, Value),
-    add_to_domain(Action, Value, Assoc, Domain_Tmp),
-    parse_parts(T, Domain_Tmp, Domain).
+    (Line = ''
```
Comment (olabknbit) [inline apps/concurrent/modules/user_input_parsing.pl:91]: This code makes the parser resistant to `\n` characters at the beginning, end and in the middle of the domain text.

File: apps/concurrent/modules/user_input_parsing.pl [hunk 3]
```diff
@@ -116,9 +119,11 @@
 parse_parts_acs([], Actions, Actions).
 parse_parts_acs([H|T], Assoc, Actions) :-
     normalize_space(atom(Line), H),
-    parse_acs_line(Line, ACS, Time),
-   	add_to_acs(ACS, Time, Assoc, Actions_Tmp),
-    parse_parts_acs(T, Actions_Tmp, Actions).
+    (Line = ''
```
Comment (olabknbit) [inline apps/concurrent/modules/user_input_parsing.pl:122]: ditto for acs

File: apps/concurrent/modules/user_input_parsing.pl [hunk 4]
```diff
@@ -136,11 +141,13 @@
 
 parse_parts_obs([], Observations, Observations).
 parse_parts_obs([H|T], Assoc, Observations) :-
-    normalize_space(string(Line), H),
-    parse_obs_line(Line, Obs, Time),
-    logic_tree_from_text(Obs, Obs_T),
-   	put_assoc(Time, Assoc, Obs_T, Observations_Tmp),
-    parse_parts_obs(T, Observations_Tmp, Observations).
+    normalize_space(atom(Line), H),
```
Comment (olabknbit) [inline apps/concurrent/modules/user_input_parsing.pl:144]: ditto for obs

**Comment by kazimierz-256** [comment]
When you try to add a release statement to an action the whole parsing predicate should fail. If you try to add already existing causes,releases,impossible the predicate should fail.

**Comment by kazimierz-256** [comment]
Ola, do you have an idea how can we handle cases where we want to assign different causes statements to the same action? Is it ok to rename the action into like say Action1 causes . if . and similarly Action2 causes . if .? If the scenario requires executing Action action should we change the scenario to require executing two different actions Action1 and Action2 and allow executing one and not the other?

**Comment by olabknbit** [comment]
> Ola, do you have an idea how can we handle cases where we want to assign different causes statements to the same action? Is it ok to rename the action into like say Action1 causes . if . and similarly Action2 causes . if .? If the scenario requires executing Action action should we change the scenario to require executing two different actions Action1 and Action2 and allow executing one and not the other?

Yeah, I was thinking about it.
Let's say we have:
```
A causes f1 and not f2 if pi
A causes g if psi
```

I see two possible solutions:
1. Currently we store an action as an assoc list of a form
`"A-["causes"-(and("f1",negate("f2")),"pi")]` for the first action (`A causes f1 and not f2 if pi`), but we can change this representation to 
`"A-["causes"-[(and("f1",negate("f2")),"pi"), ("g", "psi")]]`, for both actions, so basically the value in the assoc list for causes will be a list of pairs instead of a pair.  That would mean we would have to change how we do MNS and a lot of other things.


2. Another solution that I came across is:
We can rename the actions `A` to `A1` and `A2` respectively, but what would happen when the user actually defines actions with those names as well? We would have to double check for that. 
While parsing we would change `Scenario` so that each place where `A` is used, we put `A1` and `A2` instead. How would we handle `necessarily/possibly accessible` though? Would we ask whether `A` is fully or partially run? (E.g. `A1 and A2` or `A1 or A2`)?

This solution is safer in terms of the amount of changes we have to introduce to the existing system so I think I would go with this. 

The only thing that is slightly bothering me is:
When we have two actions like these:
```
A causes f1 and not f2 if pi or psi
A causes g if psi
```
And let's say `psi is true`, what do we do if the effects of the `A` action are conflicting? Do we treat them the same as separate actions?

We should define what is the behaviour we desire. I can later implement that.

**Comment by kazimierz-256** [comment]
I think we are opening a Pandora's box right now

**Pull request merged**